File: programmers/etc/level2/땅따먹기.py
def solution(land):
    n = len(land)
    for i in range(1, n):
        for j in range(4):
            land[i][j] = max(land[i-1][:j] + land[i-1][j+1:]) + land[i][j]
    return max(land[n-1])


land = [[1,2,3,5],[5,6,7,8],[4,3,2,1]]
print(solution(land)) # 16

File: Java/programmers/etc/level2/땅따먹기.java
package CodeTest.Java.programmers.etc.level2;

public class 땅따먹기 {
    public int solution(int[][] land) {
        int n = land.length;

        for (int i=1; i<n; i++) {
            for (int j=0; j<4; j++) {
                int tmp = 0;
                for (int k=0; k<j; k++) {
                    tmp = Math.max(tmp, land[i-1][k]);
                }
                for (int k=j+1; k<4; k++) {
                    tmp = Math.max(tmp, land[i-1][k]);
                }
                land[i][j] += tmp;
            }
        }

        int answer = 0;
        for (int j=0; j<4; j++) {
            answer = Math.max(answer, land[n-1][j]);
        }

        return answer;
    }
}
